<h1 align=center>Apply Feature Engineering Skills to the Titanic Dataset</h1>

## Description
In this exercise, we will analyze the [Titanic](https://www.kaggle.com/c/titanic) (the train subset) dataset from Kaggle.

The following is a description of the features in the data:

* **survival :** Survival (0 = No, 1 = Yes)
* **pclass :**	Ticket class (1 = 1st, 2 = 2nd, 3 = 3rd)
* **sex :**	Sex	 
* **Age :**	Age in years	 
* **sibsp :**	# of siblings / spouses aboard the Titanic	 
* **parch :**	# of parents / children aboard the Titanic	 
* **ticket :**	Ticket number	 
* **fare :** Passenger fare	 
* **cabin :**	Cabin number	 
* **embarked :**	Port of Embarkation	(C = Cherbourg, Q = Queenstown, S = Southampton)

## Import the data

In [1]:
import pandas as pd
df = pd.read_csv("./Datasets/titanic.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### Question 1:  Which feature contains the most nulls?

In [2]:
df.isnull().mean().sort_values()

PassengerId    0.000000
Survived       0.000000
Pclass         0.000000
Name           0.000000
Sex            0.000000
SibSp          0.000000
Parch          0.000000
Ticket         0.000000
Fare           0.000000
Embarked       0.002245
Age            0.198653
Cabin          0.771044
dtype: float64

*More than 77% of the Cabin feature values are null. We should be careful about imputations in this feature.*

### Question 2 : Understanding the Pclass feature

In [3]:
df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

*It's a categorical feature, so use one-hot encoding to convert to dummy variables.*

### Question 3 : Split the Fare feature into four equal bands based on the quartile boundaries.

In [4]:
# compute the quartiles values
df.describe()['Fare']

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [5]:
# create bins
Min = df.describe().loc['min','Fare']
Q1 = df.describe().loc['25%','Fare']
Q2 = df.describe().loc['50%','Fare']
Q3 = df.describe().loc['75%','Fare']
Max = df.describe().loc['max','Fare']

df['Fare_bands'] = pd.cut(df['Fare'], bins=[Min, Q1, Q2, Q3, Max], labels=['Q1', 'Q2', 'Q3', 'Q4'], include_lowest=True)
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Fare_bands  
0      0         A/5 21171   7.2500   NaN        S         Q1  
1      0          PC 17599  71.2833   C85        C         Q4  
2      0  STON/O2. 3101282   7.9250   NaN        S         Q2  
3      0            113803  53.1000  C123        S         Q4  
4      0            373450   8.0500   NaN        S         Q2

In [6]:
# check for the min and max values of Fare for each band
df.groupby('Fare_bands').agg({'Fare':['min', 'max']})

Fare          
                min       max
Fare_bands                   
Q1           0.0000    7.8958
Q2           7.9250   14.4542
Q3          14.4583   31.0000
Q4          31.2750  512.3292

### Question 4 : Examine the Age column.

In [7]:
df['Age'].sort_values().unique()

array([ 0.42,  0.67,  0.75,  0.83,  0.92,  1.  ,  2.  ,  3.  ,  4.  ,
        5.  ,  6.  ,  7.  ,  8.  ,  9.  , 10.  , 11.  , 12.  , 13.  ,
       14.  , 14.5 , 15.  , 16.  , 17.  , 18.  , 19.  , 20.  , 20.5 ,
       21.  , 22.  , 23.  , 23.5 , 24.  , 24.5 , 25.  , 26.  , 27.  ,
       28.  , 28.5 , 29.  , 30.  , 30.5 , 31.  , 32.  , 32.5 , 33.  ,
       34.  , 34.5 , 35.  , 36.  , 36.5 , 37.  , 38.  , 39.  , 40.  ,
       40.5 , 41.  , 42.  , 43.  , 44.  , 45.  , 45.5 , 46.  , 47.  ,
       48.  , 49.  , 50.  , 51.  , 52.  , 53.  , 54.  , 55.  , 55.5 ,
       56.  , 57.  , 58.  , 59.  , 60.  , 61.  , 62.  , 63.  , 64.  ,
       65.  , 66.  , 70.  , 70.5 , 71.  , 74.  , 80.  ,   nan])

In [8]:
df['Age'].isnull().value_counts()

False    714
True     177
Name: Age, dtype: int64

*A good approaches for dealing with this column :*
* *Impute null rows with the mean age across the whole dataset.*

* *Impute null rows with the mean age within each fare across different fare bands.*

### Question 5 : Which deck has the highest number of recorded passengers?
The Cabin column contains a letter and a number that corresponds to the ship's deck level. Extract the deck level to a new column called Deck.

Which deck has the highest number of recorded passengers?

In [9]:
# Check the Cabin feature values
df['Cabin'].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [10]:
# extract the first character of the Cabin
df["Deck"]  = df["Cabin"].str[0]
# compute the number of passangers for each Deck
df.Deck.value_counts()

C    59
B    47
D    33
E    32
A    15
F    13
G     4
T     1
Name: Deck, dtype: int64

*Deck C had the most recorded passengers at 59*

### Question 6 : Replace null decks with unknown.  How many unknowns do you have?

In [11]:
import numpy as np

df['Deck'] = df['Deck'].replace(np.nan, 'unknown')
df.Deck.value_counts()

unknown    687
C           59
B           47
D           33
E           32
A           15
F           13
G            4
T            1
Name: Deck, dtype: int64

*There is 687 decks with unknown values.*

### Question 7 : Which deck had the best survival rate?

In [12]:
df.groupby("Deck")["Survived"].mean().sort_values()

Deck
T          0.000000
unknown    0.299854
A          0.466667
G          0.500000
C          0.593220
F          0.615385
B          0.744681
E          0.750000
D          0.757576
Name: Survived, dtype: float64

*Deck D had the highest survival rate at 75.76%.*

### Question 8 : Which passenger classes (Pclass) were on decks A, B, and C?

In [13]:
mask = (df['Deck'] =='A')|(df['Deck'] =='B')|(df['Deck'] =='C')
df[mask]['Pclass'].unique()

array([1], dtype=int64)

*decks A, B, and C only had 1st class passengers*
### Question 9 : What would be possible actions for some of the remaining features? 

* *One-hot encode the Embarked feature.*
* *One-hot encode the Sex feature.*